<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Phylogenetic tree with trait mapping | Stephanie Chia </title> <meta name="author" content="Stephanie Chia"> <meta name="description" content="R workflow to visualize a phylogenetic tree with species-level trait data"> <meta name="keywords" content="ecoinformatics, ecology, evolution, macroecology, macroevolution, ecologist, phd, data science, environment"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link rel="preconnect" href="https://fonts.googleapis.com"> <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin> <link href="https://fonts.googleapis.com/css2?family=Klee+One&amp;family=LXGW+WenKai+TC&amp;family=Yuji+Boku&amp;display=swap" rel="stylesheet"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link defer href="/assets/css/bootstrap-toc.min.css?6f5af0bb9aab25d79b2448143cbeaa88" rel="stylesheet"> <link rel="shortcut icon" href="/assets/img/favicon.png?16a68f269ad8c4b8ec613166def673ad"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://stephchia.github.io/blog/2024/phylo-trait-plot/"> <script src="/assets/js/theme.js?a81d82887dd692e91686b43de4542f18"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav sticky-bottom-footer"> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> Stephanie Chia <span class="zh">賈媛</span> </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">About </a> </li> <li class="nav-item "> <a class="nav-link" href="/research/">Research </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">Publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/teaching/">Teaching </a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">Blog </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">CV </a> </li> <li class="nav-item"> <button id="search-toggle" title="Search" onclick="openSearchModal()"> <span class="nav-link">ctrl k <i class="ti ti-search"></i></span> </button> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="row"> <div class="col-sm-3"> <nav id="toc-sidebar" class="sticky-top"></nav> </div> <div class="col-sm-9"> <div class="post"> <header class="post-header"> <h1 class="post-title">Phylogenetic tree with trait mapping</h1> <p class="post-meta"> Created on November 30, 2024 </p> <p class="post-tags"> <a href="/blog/2024"> <i class="fa-solid fa-calendar fa-sm"></i> 2024 </a>   ·   <a href="/blog/tag/r"> <i class="fa-solid fa-hashtag fa-sm"></i> R</a>   <a href="/blog/tag/visualization"> <i class="fa-solid fa-hashtag fa-sm"></i> visualization</a>   <a href="/blog/tag/ggtree"> <i class="fa-solid fa-hashtag fa-sm"></i> ggtree</a>   <a href="/blog/tag/phylogenetics"> <i class="fa-solid fa-hashtag fa-sm"></i> phylogenetics</a>   <a href="/blog/tag/plot"> <i class="fa-solid fa-hashtag fa-sm"></i> plot</a>   ·   <a href="/blog/category/workflow"> <i class="fa-solid fa-tag fa-sm"></i> workflow</a> </p> </header> <article class="post-content"> <div id="markdown-content"> <p>==</p> <p>This note demonstrates how to visualize a phylogenetic tree with species-level trait data using R (the workflow behind Figure 4 in <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/ele.14001" rel="external nofollow noopener" target="_blank">Hung et al. 2022</a>). The tree can be plotted in either <strong>circular</strong> or <strong>linear</strong> layout formats:</p> <p><img src="/assets/img/post/20241130/01.png" alt="image" height="280"> <img src="/assets/img/post/20241130/02.png" alt="image" height="280"></p> <h2 id="1-prepare-data">1. Prepare data</h2> <h3 id="11-required-libraries">1.1 Required libraries</h3> <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggtree</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">phytools</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ggnewscale</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <h3 id="12-required-datasets">1.2 Required datasets</h3> <ul> <li>A phylogenetic tree object (<code class="language-plaintext highlighter-rouge">tree</code>) in .tre format</li> <li>A species-trait dataset (<code class="language-plaintext highlighter-rouge">trait</code>) with species names matching the tree’s tip labels</li> </ul> <h3 id="13-clean-data">1.3 Clean data</h3> <p>Reorder the trait table to match the order of species in the tree:</p> <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">is_tip</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tree</span><span class="o">$</span><span class="n">edge</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">)</span><span class="w">
</span><span class="n">ordered_tips</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tree</span><span class="o">$</span><span class="n">edge</span><span class="p">[</span><span class="n">is_tip</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">]</span><span class="w">
</span><span class="n">trait</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">trait</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="n">tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">[</span><span class="n">ordered_tips</span><span class="p">],</span><span class="w"> </span><span class="n">trait</span><span class="o">$</span><span class="n">Species</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre></div></div> <p>Set row names to species names for downstream compatibility:</p> <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">rownames</span><span class="p">(</span><span class="n">trait</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">trait</span><span class="o">$</span><span class="n">Species</span><span class="w">
</span></code></pre></div></div> <p>Example of cleaned-up species-trait data. Traits can be either continuous or categorical.</p> <table> <thead> <tr> <th>(rowname)</th> <th>Taxonomic group</th> <th>Trait 1</th> <th>Trait 2</th> <th>Trait 3</th> <th>Trait 4</th> <th>Trait 5</th> </tr> </thead> <tbody> <tr> <td>Scientific_name_1</td> <td>A</td> <td>428</td> <td>A</td> <td>C</td> <td>A</td> <td>0.09</td> </tr> <tr> <td>Scientific_name_2</td> <td>A</td> <td>274</td> <td>C</td> <td>D</td> <td>B</td> <td>0.08</td> </tr> <tr> <td>Scientific_name_3</td> <td>B</td> <td>331</td> <td>C</td> <td>C</td> <td>B</td> <td>0.24</td> </tr> <tr> <td>Scientific_name_4</td> <td>B</td> <td>189</td> <td>A</td> <td>B</td> <td>C</td> <td>0.11</td> </tr> <tr> <td>…</td> <td>…</td> <td>…</td> <td>…</td> <td>…</td> <td>…</td> <td>…</td> </tr> </tbody> </table> <p>In this tutorial, We’ll map:</p> <ul> <li> <strong>Trait 1</strong> to tree branch color</li> <li> <strong>Traits 2–5</strong> to concentric rings (using <code class="language-plaintext highlighter-rouge">gheatmap</code>)</li> <li> <strong>Group</strong> to the outermost thin ring</li> </ul> <h3 id="14-optional-estimate-ancestral-state">1.4 (Optional) Estimate ancestral state</h3> <p>If we want to color the tree using ancestral trait values, we’ll need to reconstruct internal node states. Otherwise, we can plot a single colored tree and skip this part.</p> <h4 id="141-estimate-node-values">1.4.1 Estimate node values</h4> <p>For demonstration, I use <code class="language-plaintext highlighter-rouge">phytools::fastAnc()</code> to estimate ancestral node values for a continuous trait (Trait 1). Depending on the size of the dataset, this step could be very computationally intensive.</p> <p>There are a lot of tools for ancestral state reconstruction based on specific assumptions and data format. If we already have the estimated data, just import it (and sort out how to match the data format).</p> <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">trait.1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">trait</span><span class="o">$</span><span class="n">Trait1</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">trait.1</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">trait</span><span class="p">)</span><span class="w">
</span><span class="n">fit</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fastAnc</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">trait.1</span><span class="p">)</span><span class="w"> </span><span class="c1"># computationally intensive</span><span class="w">
</span><span class="c1"># fit &lt;- readRDS("ans_fit.rds") # or import fitted data</span><span class="w">
</span></code></pre></div></div> <h4 id="142-merge-node-values-into-the-tree-data">1.4.2 Merge node values into the tree data</h4> <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">tip_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">node</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nodeid</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">trait.1</span><span class="p">)),</span><span class="w"> </span><span class="n">Trait1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trait.1</span><span class="p">)</span><span class="w">
</span><span class="n">node_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">node</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">fit</span><span class="p">)),</span><span class="w"> </span><span class="n">Trait1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">fit</span><span class="p">))</span><span class="w">
</span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">tip_data</span><span class="p">,</span><span class="w"> </span><span class="n">node_data</span><span class="p">)</span><span class="w">
</span><span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dplyr</span><span class="o">::</span><span class="n">full_join</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"node"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <h2 id="2-plotting-the-tree">2. Plotting the tree</h2> <h3 id="21-basic-tree-with-trait-1-coloring">2.1 Basic Tree with Trait 1 Coloring</h3> <p>Let’s first try out the default setting (omit <code class="language-plaintext highlighter-rouge">aes(color = Trait1)</code> if not plotting ancestral values).</p> <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ptree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggtree</span><span class="p">(</span><span class="n">tree.plot</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Trait1</span><span class="p">),</span><span class="w"> </span><span class="n">layout</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"circular"</span><span class="p">,</span><span class="w"> </span><span class="n">ladderize</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span><span class="w">
</span><span class="n">print</span><span class="p">(</span><span class="n">ptree</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p><img src="/assets/img/post/20241130/03.png" alt="image" width="70%"></p> <p>The default color scale may not reflect the full variation in Trait 1. Let’s customize the color scale for better contrast:</p> <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">palette1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">RColorBrewer</span><span class="o">::</span><span class="n">brewer.pal</span><span class="p">(</span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"RdYlBu"</span><span class="p">)</span><span class="w">
</span><span class="n">colors1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">colorRampPalette</span><span class="p">(</span><span class="n">palette1</span><span class="p">)(</span><span class="m">100</span><span class="p">)</span><span class="w">

</span><span class="n">ptree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggtree</span><span class="p">(</span><span class="n">tree.plot</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Trait1</span><span class="p">),</span><span class="w"> </span><span class="n">layout</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"circular"</span><span class="p">,</span><span class="w"> </span><span class="n">ladderize</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_colour_gradientn</span><span class="p">(</span><span class="n">colors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colors1</span><span class="p">,</span><span class="w"> </span><span class="n">trans</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"log"</span><span class="p">,</span><span class="w"> </span><span class="n">limits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">50</span><span class="p">,</span><span class="w"> </span><span class="m">500</span><span class="p">),</span><span class="w">
                         </span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">50</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">200</span><span class="p">,</span><span class="w"> </span><span class="m">400</span><span class="p">),</span><span class="w"> </span><span class="n">oob</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scales</span><span class="o">::</span><span class="n">squish</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">annotate</span><span class="p">(</span><span class="s2">"rect"</span><span class="p">,</span><span class="w"> </span><span class="n">xmin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">22</span><span class="p">,</span><span class="w"> </span><span class="n">xmax</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">35</span><span class="p">,</span><span class="w"> </span><span class="n">ymin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">-</span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="n">ymax</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.1</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">annotate</span><span class="p">(</span><span class="s2">"rect"</span><span class="p">,</span><span class="w"> </span><span class="n">xmin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">60</span><span class="p">,</span><span class="w"> </span><span class="n">xmax</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">76.5</span><span class="p">,</span><span class="w"> </span><span class="n">ymin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">-</span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="n">ymax</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.1</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w">

</span><span class="n">print</span><span class="p">(</span><span class="n">ptree</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p><img src="/assets/img/post/20241130/04.png" alt="image" width="70%"></p> <p>Key elements explained:</p> <ul> <li> <code class="language-plaintext highlighter-rouge">trans = "log"</code>: fit the right-skewed data distribution on the color scale</li> <li> <code class="language-plaintext highlighter-rouge">"limits = c(min, max)"</code>: cap the lower and upper limit of the color scale (to manage outliers)</li> <li> <code class="language-plaintext highlighter-rouge">oob = scales::squish</code>: squish extreme values into the limits (e.g. an outlier of 1,000 will be shown in the same color as 500, which is the max)</li> <li> <code class="language-plaintext highlighter-rouge">breaks</code>: specify the ticks on the legend</li> </ul> <p>A red-yellow-blue color ramp is employed using the following functions:</p> <ul> <li> <code class="language-plaintext highlighter-rouge">RColorBrewer::brewer.pal()</code>: create color palette (a vector of several colors)</li> <li> <code class="language-plaintext highlighter-rouge">colorRampPalette()</code>: extend that color palette to a ramp of 100 colors</li> </ul> <p>I also add two rings that highlight the time period of interest (in this example, the adaptive radiation events in modern bird evolution) using <code class="language-plaintext highlighter-rouge">annotation()</code>.</p> <h3 id="22-add-triat-rings-with-gheatmap">2.2 Add triat rings with <code class="language-plaintext highlighter-rouge">gheatmap()</code> </h3> <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># trait 2 (categorical)</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gheatmap</span><span class="p">(</span><span class="n">ptree</span><span class="p">,</span><span class="w"> </span><span class="n">trait</span><span class="p">[,</span><span class="w"> </span><span class="s2">"Trait2"</span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">],</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.09</span><span class="p">,</span><span class="w"> </span><span class="n">offset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-2</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">font.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">A</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#203331"</span><span class="p">,</span><span class="w"> </span><span class="n">B</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#55949A"</span><span class="p">,</span><span class="w"> </span><span class="n">C</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#C1D9B7"</span><span class="p">))</span><span class="w">

</span><span class="c1"># trait 3 (categorical)</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">new_scale_fill</span><span class="p">()</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gheatmap</span><span class="p">(</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">trait</span><span class="p">[,</span><span class="w"> </span><span class="s2">"Trait3"</span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">],</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.09</span><span class="p">,</span><span class="w"> </span><span class="n">offset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">font.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">A</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#302650"</span><span class="p">,</span><span class="w"> </span><span class="n">B</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#F3DEC0"</span><span class="p">,</span><span class="w"> </span><span class="n">C</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#DC9750"</span><span class="p">,</span><span class="w"> </span><span class="n">D</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#922C40"</span><span class="p">))</span><span class="w">

</span><span class="c1"># trait 4 (categorical)</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">new_scale_fill</span><span class="p">()</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gheatmap</span><span class="p">(</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">trait</span><span class="p">[,</span><span class="w"> </span><span class="s2">"Trait4"</span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">],</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.09</span><span class="p">,</span><span class="w"> </span><span class="n">offset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">font.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">A</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#173050"</span><span class="p">,</span><span class="w"> </span><span class="n">B</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#54A2D2"</span><span class="p">,</span><span class="w"> </span><span class="n">C</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#EBAABB"</span><span class="p">))</span><span class="w">

</span><span class="c1"># trait 5 (continuous)</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">new_scale_fill</span><span class="p">()</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gheatmap</span><span class="p">(</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">trait</span><span class="p">[,</span><span class="w"> </span><span class="s2">"Trait5"</span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">],</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.08</span><span class="p">,</span><span class="w"> </span><span class="n">offset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">28</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">font.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_gradient</span><span class="p">(</span><span class="n">low</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#ede6dd"</span><span class="p">,</span><span class="w"> </span><span class="n">high</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#4d3718"</span><span class="p">,</span><span class="w"> </span><span class="n">trans</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"log"</span><span class="p">,</span><span class="w"> </span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">.05</span><span class="p">,</span><span class="w"> </span><span class="m">.1</span><span class="p">,</span><span class="w"> </span><span class="m">.2</span><span class="p">,</span><span class="w"> </span><span class="m">.4</span><span class="p">,</span><span class="w"> </span><span class="m">.8</span><span class="p">,</span><span class="w"> </span><span class="m">1.6</span><span class="p">))</span><span class="w">

</span><span class="c1"># taxonomic group (outer thin ring)</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">new_scale_fill</span><span class="p">()</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gheatmap</span><span class="p">(</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">trait</span><span class="p">[,</span><span class="w"> </span><span class="s2">"Group"</span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">],</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.02</span><span class="p">,</span><span class="w"> </span><span class="n">offset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">43</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">font.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">P</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"gray50"</span><span class="p">,</span><span class="w"> </span><span class="n">N</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"gray80"</span><span class="p">))</span><span class="w">

</span><span class="n">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p><img src="/assets/img/post/20241130/05.png" alt="image" width="70%"></p> <p>Key elements explained:</p> <ul> <li> <code class="language-plaintext highlighter-rouge">scale_fill_manual(values = c(A = colorA, B = colorB, C = colorC))</code>: manually assign colors to each trait values. Name the color values (colorA, …) with the trait values (A, …) in the vector.</li> <li> <code class="language-plaintext highlighter-rouge">width</code>: specify the width of the ring</li> <li> <code class="language-plaintext highlighter-rouge">offset</code>: adjust the distance of the ring from the central tree</li> <li> <code class="language-plaintext highlighter-rouge">font.size = 0</code>: avoid showing labels</li> <li> <code class="language-plaintext highlighter-rouge">ggnewscale::new_scale_fill()</code>: This is required to add a new ring with a different color scheme</li> </ul> <h3 id="alternative-linear-layout-option">[Alternative] Linear layout option</h3> <p>To switch to a traditional linear tree layout, just omit <code class="language-plaintext highlighter-rouge">layout = "circular"</code>:</p> <p><img src="/assets/img/post/20241130/02.png" alt="image" width="60%"></p> <h2 id="3-final-touches">3. Final Touches</h2> <p>I often export the tree and customize the legend and annotations in PowerPoint for clarity:</p> <p><img src="/assets/img/post/20241130/01.png" alt="image" width="100%"></p> <p>We can also add silhouettes or icons to tips for visual taxonomy. Check out <a href="https://www.phylopic.org/" rel="external nofollow noopener" target="_blank">PhyloPic</a> for public domain silhouette images.</p> </div> </article> </div> </div> </div> </div> <footer class="sticky-bottom mt-5" role="contentinfo"> <div class="container"> © 2025 Stephanie Chia. Powered by Jekyll with al-folio theme. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script defer src="/assets/js/bootstrap-toc.min.js?c82ff4de8b0955d6ff14f5b05eed7eb6"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> <script type="module" src="/assets/js/search/ninja-keys.min.js?a3446f084dcaecc5f75aa1757d087dcf"></script> <ninja-keys hidebreadcrumbs noautoloadmdicons placeholder="Type to start searching"></ninja-keys> <script src="/assets/js/search-setup.js?6c304f7b1992d4b60f7a07956e52f04a"></script> <script src="/assets/js/search-data.js"></script> <script src="/assets/js/shortcut-key.js?6f508d74becd347268a7f822bca7309d"></script> </body> </html>